File: backend/app/api/v1/module_platform/tenant/permissions.py
TENANT_PERMISSION_PREFIX = "module_platform:tenant"
LEGACY_TENANT_PERMISSION_PREFIX = "module_system:tenant"

TENANT_PERMISSION_ACTIONS = ("query", "create", "update", "delete", "patch")

PLATFORM_TENANT_PERMISSIONS = {
    action: f"{TENANT_PERMISSION_PREFIX}:{action}" for action in TENANT_PERMISSION_ACTIONS
}
LEGACY_SYSTEM_TENANT_PERMISSIONS = {
    action: f"{LEGACY_TENANT_PERMISSION_PREFIX}:{action}" for action in TENANT_PERMISSION_ACTIONS
}


def tenant_permissions(action: str) -> list[str]:
    """Return primary platform permission plus legacy system alias."""
    if action not in PLATFORM_TENANT_PERMISSIONS:
        raise ValueError(f"Unsupported tenant permission action: {action}")
    return [
        PLATFORM_TENANT_PERMISSIONS[action],
        LEGACY_SYSTEM_TENANT_PERMISSIONS[action],
    ]


TENANT_QUERY_PERMISSIONS = tenant_permissions("query")
TENANT_CREATE_PERMISSIONS = tenant_permissions("create")
TENANT_UPDATE_PERMISSIONS = tenant_permissions("update")
TENANT_DELETE_PERMISSIONS = tenant_permissions("delete")
TENANT_PATCH_PERMISSIONS = tenant_permissions("patch")

File: frontend/web/src/constants/permissions.ts
const TENANT_PERMISSION_ACTIONS = ["query", "create", "update", "delete", "patch"] as const;

type TenantPermissionAction = (typeof TENANT_PERMISSION_ACTIONS)[number];

function buildTenantPermissions(prefix: "module_platform" | "module_system") {
  return TENANT_PERMISSION_ACTIONS.reduce(
    (acc, action) => {
      acc[action] = `${prefix}:tenant:${action}`;
      return acc;
    },
    {} as Record<TenantPermissionAction, string>
  );
}

export const PLATFORM_TENANT_PERMISSIONS = buildTenantPermissions("module_platform");
export const LEGACY_SYSTEM_TENANT_PERMISSIONS = buildTenantPermissions("module_system");

const PERMISSION_ALIASES = TENANT_PERMISSION_ACTIONS.reduce(
  (acc, action) => {
    const platformPermission = PLATFORM_TENANT_PERMISSIONS[action];
    const legacyPermission = LEGACY_SYSTEM_TENANT_PERMISSIONS[action];
    acc[platformPermission] = [legacyPermission];
    acc[legacyPermission] = [platformPermission];
    return acc;
  },
  {} as Record<string, string[]>
);

export function expandPermissionAliases(permission: string): string[] {
  return Array.from(new Set([permission, ...(PERMISSION_ALIASES[permission] ?? [])]));
}

export function hasPermissionInList(permissions: string[], requiredPermission: string): boolean {
  return expandPermissionAliases(requiredPermission).some((permission) =>
    permissions.includes(permission)
  );
}
